scripts/copper_bulk_analysis.py

```diff
@@ -0,0 +1,97 @@
+import matplotlib.pyplot as plt
+import numpy as np
+from ase.io import read
+
+
+# Test Convergence
+def convergence_test() -> None:
+    for n in range(1, 11):
+        atoms = read(
+            f"./data/bulk_cu_{n}x{n}x{n}/bulk_cu_{n}x{n}x{n}-000000.castep",
+        )
+        energy = atoms.get_potential_energy()
+
+        previous_energy = None
+        convergence_threshold = 1e-5  # e.g., 10^-5 eV
+        is_converged = False
+
+        if previous_energy is not None:
+            energy_diff = abs(energy - previous_energy)
+            print(f"k = {k}, Energy = {energy:.6f} eV, ΔE = {energy_diff:.6e} eV")
+
+            if energy_diff < convergence_threshold:
+                is_converged = True
+                print(f"Converged at k = {k}")
+                break
+            else:
+                print(f"k = {k}, Energy = {energy:.6f} eV")
+        previous_energy = energy
+
+        if calculation._error:
+            print("Found error in input")
+            print(calculation._error)
+
+        if not is_converged:
+            print("The energy did not converge within the tested k-point grids.")
+
+
+# Plot Potential Energy against k_points
+def potential_energy_plot() -> None:
+    potential_energies = []
+    k_points = []
+    # Load the CASTEP output file
+    for n in range(1, 11):
```

Comment:
use a generic load_bulk_calcualtion function

Reply:
don't quite understand this. I will be uploading the updated version shortly

Reply:
You should have a generic load CASTEP files from directory function which all of these use to load in the files

Reply:
Like here there is no need ro read?
```shell
for n in CASTEPConfig.k_range:
        atoms = read(f"./data/bulk_cu_{n}x{n}x{n}/bulk_cu_{n}x{n}x{n}.castep")
        energy = atoms.get_potential_energy()
```

Reply:
More like add a read all castep from folder or something - but it it doesn't make sense with other changes feel free to ignore this comment. Sometimes I suggest stuff which doesn't make sense with other changes you have made